packages/amplify-e2e-tests/src/__tests__/version-gating.test.ts

```diff
@@ -0,0 +1,232 @@
+
+  afterEach(async () => {
+    await deleteProject(projRoot);
+    deleteProjectDir(projRoot);
+  });
+
+  it('a', async () => {
+    // We cannot execute this test for packaged CLI as we are modifying the package.json file to
+    // simulate multiple versions
+    if (isPackagedAmplifyInPath()) {
+      return;
+    }
+
+    try {
+      // Reset version information to make sure tests will pass with future CLI versionsas well
+      updateVersionInPackageJson(baseCLIVersion, baseMinimumCLIVersion);
+
+      await initJSProjectWithProfile(projRoot, { name: projName });
+
+      const meta = stateManager.getMeta(projRoot);
+
+      region = _.get(meta, ['providers', 'awscloudformation', 'Region']);
+      rootStackName = _.get(meta, ['providers', 'awscloudformation', 'StackName']);
+
+      cfnClient = new CloudFormation({ region });
+
+      await expectStackMetadata(baseCLIVersion, baseMinimumCLIVersion);
+
+      // Set CLI version to 100.1.0, verify push succeeds with newer version of CLI
+      updateVersionInPackageJson('100.1.0', undefined);
+
+      await push();
+
+      await expectStackMetadata('100.1.0', baseMinimumCLIVersion);
+
+      // Set Minimum CLI version to 100.2.0, verify push succeeds with newer version of CLI
+      updateVersionInPackageJson(undefined, '100.2.0');
+
+      await pushFail();
+
+      // Set CLI version to 100.2.0, Minimum CLI version to 100.2.0, verify push succeeds
+      updateVersionInPackageJson('100.2.0', '100.2.0');
+
+      await push();
+
+      await expectStackMetadata('100.2.0', '100.2.0');
+
+      // Set CLI version to 100.3.0, Minimum CLI version to 100.2.0, verify push succeeds
+      updateVersionInPackageJson('100.3.0', '100.2.0');
+
+      await push();
+
+      await expectStackMetadata('100.3.0', '100.2.0');
+
+      // Set CLI version back to 100.2.0, Minimum CLI version to 100.2.0, verify push succeeds as
+      // minimum deploy version requirement has met
+      updateVersionInPackageJson('100.2.0', '100.2.0');
+
+      await push();
+
+      await expectStackMetadata('100.2.0', '100.2.0');
+    } finally {
+      // restore original package.json
+      fs.writeFileSync(packageJsonPath, originalPackageJsonContent);
+    }
+  });
+
+  // Test helper functions
+  const push = async () => {
+    await amplifyPushForceWithYes(projRoot);
+  };
+
+  const pushFail = async () => {
+    await amplifyPushForceWithVersionGatingOutput(projRoot);
+  };
+
+  const updateVersionInPackageJson = (cliVersion: string | undefined, minimumCLIVersion: string | undefined) => {
+    if (cliVersion) {
+      packageJson.version = cliVersion;
+    } else {
+      packageJson.version = baseCLIVersion;
+    }
+
+    if (minimumCLIVersion) {
+      packageJson.amplifyCLIConfiguration.minimumCLIVersion = minimumCLIVersion;
+    } else {
+      packageJson.amplifyCLIConfiguration.minimumCLIVersion = baseMinimumCLIVersion;
+    }
+
+    fs.writeFileSync(packageJsonPath, JSONUtilities.stringify(packageJson));
```

Comment:
It makes me a little nervous to overwrite the real package.json file. I haven't tested it, but would something like [this](https://stackoverflow.com/questions/42986480/jest-mock-import-of-json-file) with Jest mocks work here for testing purposes?

Reply:
It is an e2e test, I would not anything in e2e, overwriting the real file is the real simulation of what would be in the filesystem in that case.